ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/MessageListenerProxy.java

```diff
@@ -0,0 +1,74 @@
+package org.jboss.narayana.jta.jms;
+
+import com.arjuna.ats.jta.logging.jtaLogger;
+
+import javax.jms.Message;
+import javax.jms.MessageListener;
+import javax.jms.XASession;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.Transactional;
+import java.lang.reflect.Method;
+
+public class MessageListenerProxy implements MessageListener {
+
+    private MessageListener messageListener;
+    private TransactionManager transactionManager;
+    private XASession xaSession;
+
+    static boolean isDeclaredTransactional(MessageListener messageListener) {
+        try {
+            if (messageListener == null) {
+                return false;
+            }
+
+            Class<? extends MessageListener> listenerClass = messageListener.getClass();
+            if (listenerClass.isAnnotationPresent(Transactional.class)) {
+                return true;
+            } else {
+                Method onMessageMethod = listenerClass.getDeclaredMethod("onMessage", Message.class);
+                return onMessageMethod.isAnnotationPresent(Transactional.class);
+            }
+        } catch (NoSuchMethodException e) {
+            // Should not occur, as messageListener is an instance of javax.jms.MessageListener
+        }
+        return false;
+    }
+
+    public MessageListenerProxy(MessageListener messageListener, XASession session) {
+        this(messageListener, session, com.arjuna.ats.jta.TransactionManager.transactionManager());
+    }
+
+    /**
+     * Mainly for tests
+     */
+    MessageListenerProxy(MessageListener messageListener, XASession session, TransactionManager transactionManager ) {
+        this.messageListener = messageListener;
+        this.transactionManager = transactionManager;
+        this.xaSession = session;
+    }
+
+    @Override
+    public void onMessage(Message jmsMessage) {
```

Comment:
I asked @gytis  to look into this one. It could be a quirk of the message broker being used. Conceptually it does not seem possible for it to work (even given your experience) because the message has been removed from the queue before the enlist happened. It needs looking at in the debugger.

We could ask the Artemis team for some input if its still unclear why it is working.